### DISCUSSION TITLE
Run a matrix element only if a condition is true

### DISCUSSION THREAD

**Discussion by maximousblk** [community/community#9044] [Actions]
**Describe the enhancement**
A way to define when to run a matrix. Similar to a jobs and steps, matrices would have and `if: ...` rule

**Code Snippet**
to fit the current syntax:

```yml
runs-on: ${{ matrix.os }}
strategy:
  matrix:
    node: [12, 14, 15]
    os: [macos-latest, windows-latest, ubuntu-latest]
    include:
      - if: contains("[lts]", github.event.head_commit.message) # special value
        node: 8                                                 # regular value
        os: ubuntu-latest

      - if: contains("[lts]", github.event.head_commit.message)
        node: 10
        os: ubuntu-latest
```

if a new syntax is acceptable, I'd like to propose Matrix Groups:

```yml
runs-on: ${{ matrix.os }}
strategy:
  matrix:
    groups:
      - name: test
        with:
          os: [macos-latest, windows-latest, ubuntu-latest]
          node: [12, 14, 15]

      - name: test:lts
        if: contains("[lts]", github.event.head_commit.message)
        with:
          os: ubuntu-latest
          node: [8, 10]
```

**Additional information**

I believe this is better than having two identical jobs just with a different matrix. If there is already a good way to do this, I would really like to know.

For the second syntax, I cannot think of a way to expose the group name under the `matrix` context. For me it's fine even if it's not. And if someone really wants the name, they could always do:

```yml
matrix:
  groups:
    - name: test
      with:
        name: test
        os: [macos-latest, windows-latest, ubuntu-latest]
        node: [12, 14, 15]
```

<sub>

---
This discussion was moved here from [actions/runner#897](https://github.com/actions/runner/issues/897)

</sub>

**Comment by ethomson** [comment]
Thanks for the feedback - I definitely see the value in this.  I'd like to give some thought to more advanced workflows, like what you're describing, in the new year.  I appreciate this example.

**Reply by WoodyWoodsta** [reply]
@ethomson Any thoughts in the New Year :)

Otherwise, would like to highlight that the class of cases is also described in https://github.community/t/matrix-cannot-be-used-in-jobs-level-if/17177

I'm looking for exactly this today. My workaround is to move the conditional down into each step of the job, which isn't very elegant. Either the original feature request or the proposed ability to use `matrix` context in the `job.<name>.if` expression would solve the class of use cases.

**Comment by roehrijn** [comment]
Hi @maximousblk, @ethomson, 

may I add a bit of scope to this feature request? When I consider this to be an enhancement of `matrix` jobs to have some more attributes of a plain non-matrix job, it would definitely make sense to also implement the `needs` attribute within the matrix strategy. 

My usecase: I have a deployment pipeline with multiple Terraform stacks with interdependencies. I would like to run them in a `matrix` and just want to dynamically specify on each element which other elements (jobs) have to be run before. Currently I'm doing this by rendering the workflow.yml with a pre-processor using a commit hook but this is not very handy.

**Reply by chrispat** [reply]
Another option might be to dynamically generate the matrix https://tomasvotruba.com/blog/2020/11/16/how-to-make-dynamic-matrix-in-github-actions/

**Reply by WoodyWoodsta** [reply]
TL;DR, define two jobs: dynamically generate the matrix in the first, and pass it to the strategy in the second job definition. It's not a bad idea, but it does mean that you aren't able to get a consistent job list for the second job phase (i.e. with a false `when` clause, the outcome of the job is neutral, not non-existent).

**Comment by hoxu** [comment]
I would personally like to see `matrix` available in `jobs.foo.if`, which I guess could cover this use case as well.

**Reply by chrispat** [reply]
Right now, the `if` condition is evaluated before the matrix is expanded so the matrix context is empty at that time.  It might be possible to adjust the behavior so that we delay the evaluation of `if` when someone references the matrix as part of the condition.  I can't promise anything, but we can look into it.

**Reply by mantljosh** [reply]
FWIW I'm running into this as well and would benefit from accessing the matrix values at the job level. My use case is that I have N packages in a monorepo with expensive test suites, so I only want to start the test suite job if the package itself has changed. I can easily determine that in a separate job and use `needs`, but I need to compare the output of the "what has changed" job with the matrix value (i.e. check if this matrix variant is in the list of things that have changed)

**Reply by maximousblk** [reply]
I have since settled for a workaround that might help you:

you can create matrix ids like `packageA-16` and add them to a comma separated list as an output of a job. 
Then in the second job, in the `if` condition of the expensive/main step, I check if the current matrix id is in the list using `contains(needs.prep.outputs.should_run, '${{ matrix.package }}-${{ matrix.nodejs }}')`

**Reply by hansmbakker** [reply]
@chrispat currently, matrix jobs fail the workflow if the matrix is empty.

If you evaluate `if` only after the matrix expansion, how can we make the matrix job conditional to only run if the matrix is not empty?  

Other than that, I would like to use the matrix in the if statement

**Comment by brianjmurrell** [comment]
> Another option might be to dynamically generate the matrix https://tomasvotruba.com/blog/2020/11/16/how-to-make-dynamic-matrix-in-github-actions/

This isn't always desired.  Let's say I have a job with a matrix of 3 elements to build some software.  I then have a job with a matrix of the same 3 elements to test the build.  If one of builds fails in the first job I want to avoid trying to run testing on it which I can do with a `job.<job-id>.if` using the matrix element to access a status of the build job for the matrix element.

If I were to instead dynamically build the testing matrix to exclude the failed build element, it would indeed only run testing for the successful builds.  But if one of the builds fails due to some environmental error that I can fix and then just re-run failed jobs in the GH Actions console, even though the failed build will succeed in the re-run, it won't be tested because it would not be in the test matrix elements.

I really need to be able to fail (rather than simply exclude) the test job matrix elements so that they are eligible for being retried in a re-run.

If I could force the dynamic matrix element generation job to be run again on re-run jobs, that would I suppose work-around this issue, but since it was successful in the previous run, it will not be run on a re-run all failed jobs run.  But ultimately this is just adding more and more layers of hackery for something that should be available in workflows.
